### PR TITLE
[rush] Add support for common/temp/.rush-override

### PR DESCRIPTION
**Pull request by witcher112**
## Summary

While contributing to Rush, I've discovered that there's no easy way to test it with other repos on my PC (I was manually replacing installation in AppData/.rush directory) so I've decided to create this PR that hopefully makes it easier.

## Details

PR introduces support for common/temp/.rush-override file that can contain path to the custom rush installation. This way, contributors can override the Rush installation used by the repository and test any changes that are being made to Rush.

.rush-override shouldn't be ever tracked by git so I've decided to use common/temp directory as it should be added to .gitignore in every Rush monorepo.

If common/temp/.rush-override is present, user is presented with warning similar to the one that is displayed when using RUSH_PREVIEW_VERSION environment variable.

image

(one potential thing I would like to include here is for a warning to display the path to custom Rush installation)

If common/temp/.rush-override points to invalid Rush installation (the only check for now is verifying that package.json exists), command terminates with following error:

image

## How it was tested

I've created common/temp/.rush-override that points to custom Rush installation and executed `node apps\rush\bin\rush build --help` - seems to be working.

### CODE REVIEW

**Comment by witcher112** [comment]
I've just learned about https://rushjs.io/pages/contributing/#testing-rush-builds (right on time haha!)

But still, IMHO .rush-override could make it much easier.

**Comment by iclanton** [comment]
@dmichon-msft / @D4N14L have a potential concern about dropping this file in `common/temp`.

**Comment by witcher112** [comment]
After using it for a while, I do too - turns out `rush purge` manages to delete it (and I think it shouldn't).

Overall, I don't think it's going to be widely used that much (most probably only by contributors) to put it in already .gitignored place.

So perhaps storing it right next to rush.json should be fine, what do you think?

**Comment by iclanton** [comment]
Yeah I think putting it at the repo root is a better approach.

File: apps/rush/src/start.ts [hunk 1]
```diff
@@ -36,19 +38,77 @@ const configuration: MinimalRushConfiguration | undefined =
...
+      console.log(Colorize.red(`Cannot use common/temp/.rush-override file as it doesn't point to valid Rush package`));
+      console.log(``);
+      console.log(Colorize.red(`If you're unfamiliar with this file, you can safely delete it`));
+      process.exit(1);
+    }
+
+    const overrideVersion: string = overridePackageJson.version;
+
+    const lines: string[] = [];
+    lines.push(
```
Comment (iclanton) [inline apps/rush/src/start.ts:73]: There is a `printInBox` utility function in `@rushstack/terminal`.

File: apps/rush/src/start.ts [hunk 2]
```diff
@@ -36,19 +38,77 @@ const configuration: MinimalRushConfiguration | undefined =
...
+  version: string;
+  path: string;
+} | undefined = undefined;
+
+if (configuration) {
+  const rushOverrideFilePath: string = path.join(
+    path.dirname(configuration.rushJsonFilename),
+    rushLib.RushConstants.commonFolderName,
+    rushLib.RushConstants.rushTempFolderName,
+    '.rush-override'
```
Comment (iclanton) [inline apps/rush/src/start.ts:56]: Stick this in a constant.

File: apps/rush/src/start.ts [hunk 3]
```diff
@@ -36,19 +38,77 @@ const configuration: MinimalRushConfiguration | undefined =
...
+} | undefined = undefined;
+
+let overrideInfo: {
+  version: string;
+  path: string;
+} | undefined = undefined;
+
+if (configuration) {
+  const rushOverrideFilePath: string = path.join(
+    path.dirname(configuration.rushJsonFilename),
```
Comment (iclanton) [inline apps/rush/src/start.ts:53]:
```suggestion
    configuration.rushJsonFolder,
```

File: apps/rush/src/start.ts [hunk 4]
```diff
@@ -36,19 +38,77 @@ const configuration: MinimalRushConfiguration | undefined =
...
+    rushLib.RushConstants.commonFolderName,
+    rushLib.RushConstants.rushTempFolderName,
+    '.rush-override'
+  );
```
Comment (iclanton) [inline apps/rush/src/start.ts:57]:
```suggestion
  const rushOverrideFilePath: string = `${configuration.rushJsonFolder}/.rush-override`;
```

As per discussion.

Also stick `.rush-override` in a RushConstants.

File: apps/rush/src/start.ts [hunk 5]
```diff
@@ -36,19 +38,77 @@ const configuration: MinimalRushConfiguration | undefined =
...
+
+  if (fs.existsSync(rushOverrideFilePath)) {
```
Comment (iclanton) [inline apps/rush/src/start.ts:59]: Use `FileSystem` from `node-core-library`, and do a try-catch on reading the file and gracefully handle it not existing. See `FileSystem.isNotExistError`

File: apps/rush/src/start.ts [hunk 6]
```diff
@@ -36,19 +38,77 @@ const configuration: MinimalRushConfiguration | undefined =
...
+    const overridePath: string = fs.readFileSync(rushOverrideFilePath, 'utf8').trim();
```
Comment (iclanton) [inline apps/rush/src/start.ts:60]: Use `FileSystem.readFile`

File: apps/rush/src/start.ts [hunk 7]
```diff
@@ -36,19 +38,77 @@ const configuration: MinimalRushConfiguration | undefined =
...
+    const overridePackageJson: IPackageJson | undefined = PackageJsonLookup.instance.tryLoadPackageJsonFor(overridePath);
+
+    if (overridePackageJson === undefined) {
+      console.log(Colorize.red(`Cannot use common/temp/.rush-override file as it doesn't point to valid Rush package`));
```
Comment (iclanton) [inline apps/rush/src/start.ts:64]: Use constants here.

File: apps/rush/src/MinimalRushConfiguration.ts [hunk 8]
```diff
@@ -52,6 +54,10 @@ export class MinimalRushConfiguration {
     }
   }
 
+  public get rushJsonFilename(): string {
+    return this._rushJsonFilename;
+  }
```
Comment (iclanton) [inline apps/rush/src/MinimalRushConfiguration.ts:59]: Remove this.

**Review by iclanton — COMMENTED** [review]
What about just using an environment variable to override the rush installation/`@microsoft/rush-lib` package location?

**Comment by witcher112** [comment]
@iclanton thanks for the review!

I've decided to pursue the idea of an environment variable (I don't know why I didn't in the first place... :P)

Here's a new PR that includes rework as well as changes requested by you: https://github.com/microsoft/rushstack/pull/5003

**Pull request closed without merging**